src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1284,7 +1284,8 @@ public SolutionState WithAnalyzerReferences(IReadOnlyList<AnalyzerReference> ana
         {
             foreach (var relatedDocumentId in relatedDocumentIds)
             {
-                if (relatedDocumentId != documentId)
+                // Match the linear search behavior below and do not return documents from the same project.
+                if (relatedDocumentId != documentId && relatedDocumentId.ProjectId != documentId.ProjectId)
```

Comment:
Should we have a test for this?

Reply:
Let me take a look and see if what the testing situation for this area is